src/Shared/FileMatcher.cs

```diff
@@ -836,6 +837,20 @@ struct RecursionState
             Dictionary<string, List<RecursionState>> searchesToExcludeInSubdirs,
             TaskOptions taskOptions)
         {
+#if NET6_0
+            // This is a pretty quick, simple check, but it misses some cases:
+            // symlink in folder A pointing to folder B and symlink in folder B pointing to folder A
+            // GetFiles in folder C that contains a symlink that points folder D, which contains C as well as other files
```

Comment:
"If folder C contains file Foo.cs and folder D, and folder D contains a symlink pointing to folder C, calling GetFilesRecursive and passing in folder D would currently find Foo.cs, whereas this would make us miss it."

Is that a clearer description?

That sounds like a valid case, but I feel fine about ignoring it for now, since it would hang with current MSBuild anyway, so we know no one is doing it.

Reply:
Ah, I see. Could you avoid the recursion _into_ the "bad" symlink, instead of just returning from it quickly? That feels like the same basic code, but would solve that case (which I think is important).

Reply:
How would that help? Either we follow the symlink or we don't, and if we do, then we can only avoid following it repeatedly if we store the visited directories; if we don't, we can't possibly find Foo.cs in that case.

Reply:
No; with my approach, that scenario would work. It's fairly easy to modify my test to prove that. The scenario that wouldn't work is:

* C
  * D
    * AAA (link to C)
  * Foo.cs

The idea then would be that it would start at D, check AAA, and not follow it, and be finished. Ideally, it would start at D, follow AAA to C, find Foo.cs, check D again, see AAA again but _not_ follow it this time, and stop. That ideal isn't possible now, though, which is why I think this solution is fine.